<a href="https://colab.research.google.com/github/theysee/sberbank_reboot/blob/master/%D0%9F%D0%BE%D1%82%D0%BE%D0%BA_3_%D0%9F%D1%80%D0%B0%D0%BA%D1%82%D0%B8%D0%BA%D0%B0_%D0%BA_%D0%B7%D0%B0%D0%BD%D1%8F%D1%82%D0%B8%D1%8E_5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Практика



*   [Презентация](https://docs.google.com/presentation/d/1dLjCBWeGlhzU3A_JNg09ty-WPiAe4XObZBENp4tgC2Y/edit?usp=sharing)
*   [master](https://colab.research.google.com/drive/1TVa3TRpGfhQa3f6eKxAxmBEHsPHUVzxt)
*   [Дамп базы](https://drive.google.com/open?id=1RJx-1kh8az1OSQqnsDjCTWdzJBlDkjcB)



![схема базы](http://bdis.umeta.ru/db/db_course/labs/SQLpHTM/king.gif)

In [62]:
!pip install PyMySQL

In [63]:
!pip install pandas

In [0]:
import pymysql
import pandas as pd
# доступны student-ы от 0 до 49

DB_HOST = '89.223.95.235'
DB_USER = 'student_12'
DB_USER_PASSWORD = 'student_12'
DB_NAME = 'sample_db'

In [79]:
conn = pymysql.connect(DB_HOST, DB_USER, DB_USER_PASSWORD, DB_NAME)
pd.read_sql_query("select * from employee", conn).head(10)

employee_id last_name first_name  ...  salary  commission  department_id
0         7369     smith       john  ...   800.0         NaN             20
1         7499     allen      kevin  ...  1600.0       300.0             30
2         7505     doyle       jean  ...  2850.0         NaN             13
3         7506    dennis       lynn  ...  2750.0         NaN             23
4         7507     baker     leslie  ...  2200.0         NaN             14
5         7521      ward    cynthia  ...  1250.0       500.0             30
6         7555    peters     daniel  ...  1250.0       300.0             13
7         7557      shaw      karen  ...  1250.0      1200.0             13
8         7560    duncan      sarah  ...  1250.0         NaN             23
9         7564     lange    gregory  ...  1250.0       300.0             23

[10 rows x 10 columns]

### Задание 1
Выбрать всю информацию о каждом отделе и месте его расположения.

In [66]:
pd.read_sql_query("""
  SELECT
      department_id,
      name,
      regional_group,
      location.location_id
  FROM
      department
      INNER JOIN
      location
      ON department.location_id = location.location_id
""", conn)

department_id        name regional_group  location_id
0              10  accounting       new york          122
1              12    research       new york          122
2              13       sales       new york          122
3              14  operations       new york          122
4              20    research         dallas          124
5              23       sales         dallas          124
6              24  operations         dallas          124
7              30       sales        chicago          123
8              34  operations        chicago          123
9              40  operations         boston          167
10             43       sales         boston          167

### Задание 2
Выбрать список фамилий сотрудников по городам, где они работают, с указанием для каждого его должности.

In [67]:
pd.read_sql_query("""
  SELECT
    employee_id,
    function 'должность',
    name as 'Имя отдела',
    regional_group 'город'
  FROM
    employee
  INNER JOIN
    job
  INNER JOIN
    department
  INNER JOIN
    location
  ON employee.job_id = job.job_id AND employee.department_id = department.department_id AND department.location_id = location.location_id
  ORDER BY
    'Имя отдела'
""", conn)

employee_id    должность  Имя отдела     город
0          7782      manager  accounting  new york
1          7839    president  accounting  new york
2          7934        clerk  accounting  new york
3          7569      manager    research  new york
4          7799      analyst    research  new york
5          7916      analyst    research  new york
6          7919        clerk    research  new york
7          7505      manager       sales  new york
8          7555  salesperson       sales  new york
9          7557  salesperson       sales  new york
10         7600  salesperson       sales  new york
11         7950        clerk       sales  new york
12         7507      manager  operations  new york
13         7369        clerk    research    dallas
14         7566      manager    research    dallas
15         7788      analyst    research    dallas
16         7876        clerk    research    dallas
17         7902      analyst    research    dallas
18         7506      manager       sales    dallas
19         7560  salesperson       sales    dallas
20         7564  salesperson       sales    dallas
21         7789  salesperson       sales    dallas
22         7954        clerk       sales    dallas
23         7609        staff  operations    dallas
24         7499  salesperson       sales   chicago
25         7521  salesperson       sales   chicago
26         7654  salesperson       sales   chicago
27         7698      manager       sales   chicago
28         7844  salesperson       sales   chicago
29         7900        clerk       sales   chicago
30         7676        staff  operations   chicago
31         7820  salesperson       sales    boston

### Задание 3
 Выбрать всех покупателей, с которыми работает продавец TURNER.

In [68]:
pd.read_sql_query('''
  SELECT
    customer.name
  FROM
    employee
    INNER JOIN
    customer
    ON employee.employee_id = customer.salesperson_id AND employee.last_name = 'TURNER'
''', conn)

name
0                                    jocksports
1                                  k + t sports
2  north woods health and fitness supply center

### Задание 4
Выбрать сотрудников, которые не закреплены за конкретным заказчиком.

In [69]:
pd.read_sql_query('''
  SELECT
    employee.first_name,
    employee.last_name,
    customer.name
  FROM
    employee
    LEFT OUTER JOIN
    customer
    ON employee.employee_id=customer.salesperson_id AND name IS NULL
''', conn)

first_name last_name  name
0        john     smith  None
1       kevin     allen  None
2        jean     doyle  None
3        lynn    dennis  None
4      leslie     baker  None
5     cynthia      ward  None
6      daniel    peters  None
7       karen      shaw  None
8       sarah    duncan  None
9     gregory     lange  None
10      terry     jones  None
11      chris   alberts  None
12    raymond    porter  None
13    richard     lewis  None
14    kenneth    martin  None
15     denise   sommers  None
16     marion     blake  None
17      carol     clark  None
18     donald     scott  None
19      livia      west  None
20    matthew    fisher  None
21       paul      ross  None
22    francis      king  None
23       mary    turner  None
24      diane     adams  None
25       fred     james  None
26   jennifer      ford  None
27      grace   roberts  None
28    michael   douglas  None
29    barbara    miller  None
30      alice    jensen  None
31      james    murray  None

### Задание 5
Выбрать имена продавцов, которые работают более чем с одним покупателем и названия покупателей, которых они обслуживают.

In [70]:
pd.read_sql_query('''
  SELECT
    employee.first_name,
    employee.last_name,
    COUNT(customer.name) AS 'Count'
  FROM
    employee,
    customer
  WHERE
    employee.employee_id=customer.salesperson_id
  GROUP BY
    employee.first_name,
    employee.last_name
  ORDER BY
    employee.last_name
''', conn)

first_name last_name  Count
0       kevin     allen      2
1       sarah    duncan      2
2     gregory     lange      2
3     kenneth    martin      2
4      daniel    peters      2
5     raymond    porter      2
6        paul      ross      7
7       karen      shaw      4
8        mary    turner      3
9     cynthia      ward      3
10      livia      west      4

In [71]:
pd.read_sql_query('''
  SELECT
    employee.first_name,
    employee.last_name,
    customer.name
  FROM
    employee,
    customer
  WHERE
    employee.employee_id=customer.salesperson_id
  ORDER BY
    employee.last_name
''', conn)

first_name last_name                                          name
0       kevin     allen                                 womens sports
1       kevin     allen                                every mountain
2       sarah    duncan                             the power forward
3       sarah    duncan                           al and bob's sports
4     gregory     lange                           hit, throw, and run
5     gregory     lange                                 al's pro shop
6     kenneth    martin                           bob's family sports
7     kenneth    martin                                     vollyrite
8      daniel    peters                                 fitness first
9      daniel    peters                                  century shop
10    raymond    porter                                    good sport
11    raymond    porter                                     all sport
12       paul      ross                              the all american
13       paul      ross                                         hoops
14       paul      ross                                        at bat
15       paul      ross                                   velo sports
16       paul      ross                                  the outfield
17       paul      ross                                    fast break
18       paul      ross                                      the tour
19      karen      shaw                                stadium sports
20      karen      shaw                                   point guard
21      karen      shaw                                  the coliseum
22      karen      shaw                                rebound sports
23       mary    turner                                    jocksports
24       mary    turner  north woods health and fitness supply center
25       mary    turner                                  k + t sports
26    cynthia      ward                                      shape up
27    cynthia      ward                                   just tennis
28    cynthia      ward                                tkb sport shop
29      livia      west                               joe's bike shop
30      livia      west                              wheels and deals
31      livia      west                    bob's swim, cycle, and run
32      livia      west                                    just bikes

### Задание 6
Выбрать сумму всех продаж, которые обеспечил продавец TURNER.

In [72]:
pd.read_sql_query('''
  SELECT
    employee.first_name,
    employee.last_name,
    SUM(sales_order.total)
  FROM
    employee,
    sales_order
  WHERE
    last_name = 'turner'
  GROUP BY
    employee.first_name,
    employee.last_name
''', conn)

first_name last_name  SUM(sales_order.total)
0       mary    turner               237236.75

### Задание 7
Выбрать сумму всех продаж товаров, связанных с теннисом, за лето 1990г.

In [73]:
pd.read_sql_query('''
  SELECT
    product.description,
    sales_order.order_date
  FROM
    product,
    sales_order,
    item
  WHERE
    sales_order.order_date > '1990-06-01' 
    AND 
    sales_order.order_date < '1990-09-01'
    AND 
    product.description LIKE '%tennis%'
''', conn)

description order_date
0          ace tennis racket i 1990-06-14
1         ace tennis racket ii 1990-06-14
2      ace tennis balls-3 pack 1990-06-14
3      ace tennis balls-6 pack 1990-06-14
4               ace tennis net 1990-06-14
...                        ...        ...
26553  ace tennis balls-3 pack 1990-08-01
26554  ace tennis balls-6 pack 1990-08-01
26555           ace tennis net 1990-08-01
26556         sp tennis racket 1990-08-01
26557    rh: "guide to tennis" 1990-08-01

[26558 rows x 2 columns]

### Задание 8
Выбрать имена и зарплаты самого старого и самого молодого сотрудника фирмы. ("Возраст" сотрудника определять по дате его поступления на работу в фирму).

In [74]:
pd.read_sql_query('''
  SELECT
    first_name,
    last_name,
    salary,
    hire_date
  FROM
    employee
  WHERE
    hire_date IN (SELECT MIN(hire_date) FROM employee)
    OR
    hire_date IN (SELECT MAX(hire_date) FROM employee)
''', conn)

first_name last_name  salary  hire_date
0       john     smith   800.0 1984-12-17
1      james    murray   750.0 1987-01-16

### Задание 9
Для каждого продавца вывести разность между его зарплатой и средней зарплатой продавцов в фирме. Решение

In [75]:
pd.read_sql_query('''
  SELECT
    first_name,
    last_name,
    salary,
    salary - (SELECT AVG(salary) FROM employee) AS Calculate
  FROM
    employee
''', conn)

first_name last_name  salary  Calculate
0        john     smith   800.0  -1096.875
1       kevin     allen  1600.0   -296.875
2        jean     doyle  2850.0    953.125
3        lynn    dennis  2750.0    853.125
4      leslie     baker  2200.0    303.125
5     cynthia      ward  1250.0   -646.875
6      daniel    peters  1250.0   -646.875
7       karen      shaw  1250.0   -646.875
8       sarah    duncan  1250.0   -646.875
9     gregory     lange  1250.0   -646.875
10      terry     jones  2975.0   1078.125
11      chris   alberts  3000.0   1103.125
12    raymond    porter  1250.0   -646.875
13    richard     lewis  1800.0    -96.875
14    kenneth    martin  1250.0   -646.875
15     denise   sommers  1850.0    -46.875
16     marion     blake  2850.0    953.125
17      carol     clark  2450.0    553.125
18     donald     scott  3000.0   1103.125
19      livia      west  1500.0   -396.875
20    matthew    fisher  3000.0   1103.125
21       paul      ross  1300.0   -596.875
22    francis      king  5000.0   3103.125
23       mary    turner  1500.0   -396.875
24      diane     adams  1100.0   -796.875
25       fred     james   950.0   -946.875
26   jennifer      ford  3000.0   1103.125
27      grace   roberts  2875.0    978.125
28    michael   douglas   800.0  -1096.875
29    barbara    miller  1300.0   -596.875
30      alice    jensen   750.0  -1146.875
31      james    murray   750.0  -1146.875

### Задание 10
Вывести имена, средние инициалы и фамилии всех однофамильцев.

In [76]:
pd.read_sql_query('''
  SELECT
    employee.first_name,
    employee.last_name,
    department.name
  FROM
    department
  INNER JOIN
    employee
  ON department.department_id = employee.department_id
  ORDER BY
    department.name
''', conn)
#Однофамильцев нет.

first_name last_name        name
0       carol     clark  accounting
1     francis      king  accounting
2     barbara    miller  accounting
3      leslie     baker  operations
4     richard     lewis  operations
5      denise   sommers  operations
6       chris   alberts    research
7     matthew    fisher    research
8       grace   roberts    research
9     michael   douglas    research
10       john     smith    research
11      terry     jones    research
12     donald     scott    research
13      diane     adams    research
14   jennifer      ford    research
15       jean     doyle       sales
16     daniel    peters       sales
17      karen      shaw       sales
18    raymond    porter       sales
19      alice    jensen       sales
20       lynn    dennis       sales
21      sarah    duncan       sales
22    gregory     lange       sales
23      livia      west       sales
24      james    murray       sales
25      kevin     allen       sales
26    cynthia      ward       sales
27    kenneth    martin       sales
28     marion     blake       sales
29       mary    turner       sales
30       fred     james       sales
31       paul      ross       sales

### Задание 11
Для каждого отдела выбрать процент, который составляют сотрудники отдела от общего числа сотрудников фирмы, и процент, который составляет зарплата отдела от общей зарплаты фирмы.

In [78]:
pd.read_sql_query('''
  SELECT
    (COUNT(employee.employee_id)/(SELECT COUNT(employee_id) AS 'Count' FROM employee GROUP BY 'Count'))*100 AS 'Percent_Count',
    (SUM(employee.salary)/(SELECT SUM(salary) AS 'Sum' FROM employee GROUP BY 'Sum'))*100 AS 'Percent_Sum',
    department.name
  FROM
    department
  INNER JOIN
    employee
  ON department.department_id = employee.department_id
  GROUP BY
    department.name
  ORDER BY
    department.name
''', conn)

Percent_Count  Percent_Sum        name
0          9.375    14.415157  accounting
1          9.375     9.637562  operations
2         28.125    33.855025    research
3         53.125    42.092257       sales

### Задание 12
Выбрать фамилии начальников тех отделов, которые имеют разветвленную структуру (т.е. тех начальников, у прямых подчиненных которых есть свои подчиненные).

In [98]:
pd.read_sql_query('''
  SELECT
    first_name,
    last_name,
    employee_id,
    manager_id
  FROM
    employee
  WHERE
    employee_id IN (SELECT manager_id FROM employee WHERE employee_id IN (SELECT manager_id FROM employee GROUP BY manager_id) GROUP BY manager_id)
''', conn)

first_name last_name  employee_id  manager_id
0      terry     jones         7566      7839.0
1      chris   alberts         7569      7839.0
2    francis      king         7839         NaN

### Задание 13
Выбрать названия отдела, заключившего сделки на самую большую сумму, и города, в котором он находится.

### Задание 14
Выбрать название товара, дату продажи, цену продажи для всех случаев, когда товары продавались ниже, чем за 75% их объявленной цены.

### Задание 15
Выбрать общую сумму скидок, предоставленных покупателю STADIUM SPORTS в 1989г.

### Задание 16
Выбрать список сотрудников фирмы с указанием фамилии непосредственного начальника каждого.

### Задание 17
 Для каждого сотрудника вывести разность между его зарплатой и средней зарплатой сотрудников, выполняющих те же функции.

### Задание 18
Вывести таблицу распределения объема продаж товара DUNK BASKETBALL INDOOR по годам.



### Задание 19
Выбрать фамилии тех сотрудников, у которых суммарный доход (зарплата + комиссионные) больше 2000.

### Задание 20
Выбрать названия товаров, для которых нынешняя цена увеличилась по сравнению с ценой на 15 декабря 1989г. более, чем на 15%.